### ISSUE TITLE
Support variables in env.yaml

### ISSUE THREAD

**Issue by williewheeler-ms** [ploomber/ploomber#696]
Compare:

```
data_out_dir = "../output/data"
models_out_dir = "../output/models"
alerts_out_dir = "../output/alerts"
```

vs.

```
out_dir = "../output"
data_out_dir = "{{out_dir}}/data"
models_out_dir = "{{out_dir}}/models"
alerts_out_dir = "{{out_dir}}/alerts"
```

**Comment by edublancas** [comment]
Interesting, are you thinking of something flexible?

```yaml
a: value
b: '{{a}}/something'
c: '{{b}}/another'
```

Or have a separate section:

```yaml
variables:
    a: value

b: '{{a}}/something'
c: '{{a}}/something/another'
```

The first one is succinct, but for large `env.yaml` files - it might get a bit hard to read. The second one is more strict, but the structure is clear. Thoughts?

**Comment by williewheeler-ms** [comment]
```
     __
    /,,\o
    ____
  o/,__,\
   ______
  /,____,\o
  ________
o/,______,\
 __________
/,________,\o
```

https://codegolf.stackexchange.com/questions/87614/turtles-all-the-way-down

**Comment by edublancas** [comment]
Yeah makes sense. Thanks for the feedback!

So this one is gonna be interesting because we need to represent the `env.yaml` values as a DAG, then sort topologically and then render. [Here's where it should happen.](https://github.com/ploomber/ploomber/blob/f2b44e15db57e372a97ca8fa61d1f728fd659714/src/ploomber/env/envdict.py#L101) If you wanna give a try, feel free to!

**Comment by williewheeler-ms** [comment]
I don't know that it requires a topological sort. Could just process the file line by line. Most users would expect the variable to be defined before being used.

**Comment by edublancas** [comment]
Yeah, that's true.

I guess it can be implemented serially and just provide useful error messages like "variable A hasn't been defined yet, but it is defined in line X" - just to hint to the user that they might need to move it

**Comment by williewheeler-ms** [comment]
Yes. In my view that would be preferable, and not simply from an implementation perspective. The mental model is just scripty variable assignment, and so users would expect order to matter.

Also allows you to sidestep multiple definitions, which would be a problem under topological sort. It might make sense to disallow that with the serial approach anyway, since it seems more likely than not that a reassignment would be a mistake in this kind of file. But you could allow it if you like, and serial processing would work unambiguously.

**Comment by williewheeler-ms** [comment]
Oh, and to your suggested error message -- sure, that would be great. Something I really appreciate about ploomber is that you do a good job of trying to figure out what the user was trying to do and offer hints. I see your idea in that light.

**Comment by edublancas** [comment]
Sounds good, I think the serial approach is best as long as there are good hints for users (throw warnings/errors on re-definitions, helpful error messages if variables are used before they're defined).

**Comment by edublancas** [comment]
closed by 8a9fccd14938d805941046d2d65d6cd6a8d9468e

**Comment by eromoe** [comment]
There is a problem if you change `out_dir`  by  `--env--out_dir  ../another/place` , others dependencies would remain old value.
```
out_dir = "../output"
data_out_dir = "{{out_dir}}/data"
models_out_dir = "{{out_dir}}/models"
alerts_out_dir = "{{out_dir}}/alerts"
```
became 
```
out_dir = "../another/place"
data_out_dir = "../output/data"
models_out_dir = "../output/models"
alerts_out_dir = "../output/alerts"
```

I'd like change on parent variable would derive to the children.
